File: app/src/main/java/meugeninua/appfromscratch/app/managers/AppEventsManager.java
package meugeninua.appfromscratch.app.managers;

import android.os.Handler;
import android.os.Looper;
import android.support.annotation.AnyThread;
import android.support.annotation.MainThread;
import android.support.v4.util.ArrayMap;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collection;
import java.util.List;
import java.util.Map;
import java.util.UUID;

import javax.inject.Inject;

import meugeninua.appfromscratch.app.di.scopes.PerApplication;
import timber.log.Timber;

@PerApplication
public class AppEventsManager {

    private final Map<UUID, ObserverWrapper<?>> observers;
    final Handler handler;

    @Inject
    AppEventsManager() {
        observers = new ArrayMap<>();
        handler = new Handler(Looper.getMainLooper());
    }

    @AnyThread
    public void post(final Object event) {
        Timber.d(Thread.currentThread().getName());

        final List<ObserverWrapper<?>> wrappers;
        synchronized (this) {
            wrappers = new ArrayList<>(observers.values());
        }
        for (ObserverWrapper<?> wrapper : wrappers) {
            wrapper.update(this, event);
        }
    }

    @AnyThread
    public <T> UUID subscribeToEvent(
            final Class<T> clazz,
            final TypedObserver<T> observer) {
        UUID key = null;
        synchronized (this) {
            while (key == null || observers.containsKey(key)) {
                key = UUID.randomUUID();
            }
            final ObserverWrapper<T> impl = new ObserverWrapper<>(
                    clazz, observer);
            observers.put(key, impl);
        }
        return key;
    }

    @AnyThread
    public void unsubscribe(final UUID... keys) {
        unsubscribe(Arrays.asList(keys));
    }

    @AnyThread
    public synchronized void unsubscribe(final Collection<UUID> keys) {
        for (UUID key : keys) {
            Timber.d("Unsubscribed %s", key);
            observers.remove(key);
        }
    }

    private static class ObserverWrapper<T> {

        private final Class<T> clazz;
        private final TypedObserver<T> observer;

        ObserverWrapper(
                final Class<T> clazz,
                final TypedObserver<T> observer) {
            this.clazz = clazz;
            this.observer = observer;
        }

        @AnyThread
        void update(
                final AppEventsManager manager,
                final Object arg) {
            if (clazz.isInstance(arg)) {
                manager.handler.post(() -> observer.onUpdate(clazz.cast(arg)));
            }
        }
    }

    public interface TypedObserver<T> {

        @MainThread
        void onUpdate(T event);
    }
}

File: app/src/main/java/meugeninua/appfromscratch/app/managers/AppEventsManager.kt
package meugeninua.appfromscratch.app.managers

import android.arch.lifecycle.GenericLifecycleObserver
import android.arch.lifecycle.Lifecycle
import android.arch.lifecycle.LifecycleOwner
import android.os.Handler
import android.os.Looper
import android.support.annotation.AnyThread
import android.support.annotation.MainThread
import android.support.v4.util.ArrayMap

import java.lang.ref.WeakReference
import java.util.ArrayList
import java.util.UUID

import javax.inject.Inject
import javax.inject.Singleton

import timber.log.Timber

@Singleton
class AppEventsManager @Inject internal constructor() {

    private val observers: MutableMap<UUID, ObserverWrapper<*>> = ArrayMap()
    internal val handler: Handler = Handler(Looper.getMainLooper())

    @AnyThread
    fun post(event: Any) {
        Timber.d(Thread.currentThread().name)

        lateinit var wrappers: List<ObserverWrapper<*>>
        synchronized(this) {
            wrappers = ArrayList(observers.values)
        }
        for (wrapper in wrappers) {
            wrapper.update(this, event)
        }
    }

    @MainThread
    fun <T> subscribeToEvent(
            owner: LifecycleOwner,
            clazz: Class<T>,
            observer: (T) -> Unit) {
        val impl = ObserverWrapper<T>()
        impl.observer = observer
        impl.clazz = clazz
        impl.ownerRef = WeakReference(owner)
        impl.managerRef = WeakReference(this)

        var key: UUID = UUID.randomUUID()
        synchronized(this) {
            while (observers.containsKey(key)) {
                key = UUID.randomUUID()
            }
            impl.key = key
            observers.put(key, impl)
        }
        impl.attachToLifecycle()
    }

    @MainThread
    @Synchronized
    internal fun unsubscribe(key: UUID?) {
        Timber.d("Unsubscribed %s", key)
        val wrapper = observers.remove(key)
        wrapper?.detachFromLifecycle()
    }

    private class ObserverWrapper<T> : GenericLifecycleObserver {

        lateinit var clazz: Class<T>
        lateinit var observer: (T) -> Unit
        lateinit var key: UUID
        lateinit var ownerRef: WeakReference<LifecycleOwner>
        lateinit var managerRef: WeakReference<AppEventsManager>

        override fun onStateChanged(
                source: LifecycleOwner,
                event: Lifecycle.Event) {
            val manager = managerRef.get()
            if (manager != null && source.lifecycle
                            .currentState == Lifecycle.State.DESTROYED) {
                manager.unsubscribe(key)
            }
        }

        internal fun attachToLifecycle() {
            val owner = ownerRef.get() ?: return
            owner.lifecycle.addObserver(this)
        }

        @MainThread
        internal fun detachFromLifecycle() {
            val owner = ownerRef.get() ?: return
            owner.lifecycle.removeObserver(this)
        }

        @AnyThread
        internal fun update(
                manager: AppEventsManager,
                arg: Any) {
            if (clazz.isInstance(arg)) {
                manager.handler.post { observer(clazz.cast(arg)) }
            }
        }
    }
}
